File: 2473-max-sum-of-a-pair-with-equal-sum-of-digits/2473-max-sum-of-a-pair-with-equal-sum-of-digits.java
class Solution {

    public static int sumOfDigits(int num) {
        int sum = 0;
        while (num > 0) {
            sum += num % 10;
            num /= 10;
        }
        return sum;
    }

    public int maximumSum(int[] nums) {
        Map<Integer, int[]> map = new HashMap<>();
        int maxSum = -1;

        for (int num : nums) {
            int digitSum = sumOfDigits(num);

            // If key is absent, initialize an array to store top 2 numbers
            map.putIfAbsent(digitSum, new int[]{-1, -1});

            int[] topTwo = map.get(digitSum);

         
            if (num > topTwo[0]) {
                topTwo[1] = topTwo[0];
                topTwo[0] = num;      
            } else if (num > topTwo[1]) {
                topTwo[1] = num;      
            }

           
            if (topTwo[1] != -1) {
                maxSum = Math.max(maxSum, topTwo[0] + topTwo[1]);
            }
        }

        return maxSum;
    }}

File: 2473-max-sum-of-a-pair-with-equal-sum-of-digits/2473-max-sum-of-a-pair-with-equal-sum-of-digits.cpp
class Solution {
public:
    int digitSum(int n) {
        int res = 0;
        int num = n;
        while (num > 0) {
            res += num % 10;
            num /= 10;
        }
        return res;
    }
    int maximumSum(vector<int>& nums) {
        map<int, pair<int, int>> mp;

        int res = -1;

        for (int num : nums) {

            int dSum = digitSum(num);

            auto& p = mp[dSum];

            
            if (num > p.first) {
                p.second = p.first;
                p.first = num;

            } else if (num > p.second) {
                p.second = num;
            }

            if(p.second!=0){
                res=max(res,p.first+p.second);
            }

           // cout<<num<<" "<<p.first<<" "<<p.second<<endl;
        }
        return res;
    }
};